certora/specs/Delegation.spec

```
@@ -0,0 +1,61 @@
+import "ERC20.spec";
+
+ghost mathint sumOfVotingPower {
+    init_state axiom sumOfVotingPower == 0;
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sload uint256 votingPower (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] {
+    require sumOfVotingPower >= to_mathint(votingPower);
+}
+
+// Slot for DelegationTokenStorage._delegatedVotingPower
+hook Sstore (slot 0x669be2f4ee1b0b5f3858e4135f31064efe8fa923b09bf21bf538f64f2c3e1101)[KEY address addr] uint256 newValue (uint256 oldValue) {
+    sumOfVotingPower = sumOfVotingPower - oldValue + newValue;
+}
+
+// Check that zero address has no voting power assuming that zero address can't make transactions.
+invariant zeroAddressNoVotingPower()
+    delegatee(0x0) == 0x0 && delegatedVotingPower(0x0) == 0
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    { preserved with (env e) { require e.msg.sender != 0; } }
+
+// Check that the voting power is never greater than the total supply of tokens.
+invariant totalSupplyLTEqSumOfVotingPower()
+    to_mathint(totalSupply()) == sumOfVotingPower + currentContract._zeroVirtualVotingPower
+    filtered {
+      // Ignore upgrades.
+      f-> f.selector != sig:upgradeToAndCall(address, bytes).selector
+    }
+    {
+      preserved  with (env e) {
+          requireInvariant totalSupplyIsSumOfBalances();
+          requireInvariant zeroAddressNoVotingPower();
+      }
+    }
+
+// Check that user can restore their voting power by delegating to zero address then delgating back to themselves.
+rule delgatingSelfConsistent {
```

```suggestion
rule delegateToSelf {
```
"is consistent" or "is correct" feels very generic, I think we can avoid this phrasing

---

We can keep "consistent", but just using "consistent" or "correct" feels awkward. One of the goals of the formal verification is to explicit what it means to be consistent or correct. So I think we should be more specific

---

Good, I couldn't agree more!

---

Changed in https://github.com/morpho-org/morpho-token-upgradeable/pull/92/commits/f80c7fbaebcbaf0f9b92f71aeafeed59a04c15fc.